### DISCUSSION TITLE
How not to expose classes from indirectly included files?

### DISCUSSION THREAD

**Discussion by jwuttke** [RosettaCommons/binder#183] [Q&A] [answered]
Do I understand correctly?:

- binder generates pybind11 macros for every class, struct, function, constant declared in the header files supplied as command-line argument, and in the files included by these files, and in files recursively included to any recursion depth?
- to avoid deep recursions, it is advisable to supply flat header files that directly include all files which otherwise would be included indirectly.

Assuming that this is correct, then this is a big disadvantage with respect to Swig, where there is a distinction between `#include` (necessary to digest the input code) and `%include` (listing files that should be exposed to the target language).

Or is there a way to prevent binder from generating pybind11 macros for classes etc in indirectly included files?

**Comment by lyskov** [comment, marked as answer]
Hm... not exactly, let me clarify Binder workflow a bit:
- all-includes.hpp is file that should directly or indirectly provide provide definition to all types/functions that we want to expose in Python. It does not matter if files included directly or through other includes, there should not be any difference.
- Binder uses config options to control what exactly should be bound and what should be skipped (types/functions that Binder think could not be bound will be skipped automatically). Only these config options influence set of bindings that will be generated. Please see https://cppbinder.readthedocs.io/en/latest/config.html for list of config options. Particularly `+-namespace`, `+-class`.

@jwuttke does this answer your question?

**Reply by jwuttke** [reply]
Thank you. Under https://cppbinder.readthedocs.io/en/latest/config.html > Config file options > Config file directives, there is also a `+-include` directive. I am unsure though what is meant by "adding to generated file" - I guess it's about adding "#include" statements, not about generating bindings for the content of the include files?

So there is no way to steer the generation of pybind11 macros at the level of files instead of namespaces or classes?

**Reply by lyskov** [reply]
re `+-include`: yes, this directives will include specified header file in generated code. This is necessary when you want to provide custom binding functions for some types (for example for templates classes).

> So there is no way to steer the generation of pybind11 macros at the level of files instead of namespaces or classes?

-- currently this is not possible. If you want to avoid bindings generation for particular parts of the code the only way would be either to not add such files into all-includes.hpp or explicitly disable classes/namespaces.
